### DISCUSSION TITLE
Semantic File Inspector interoperability

### DISCUSSION THREAD

**Discussion by IS4Code** [SPARQL-Anything/sparql.anything#470] [Ideas]
A little introduction: [_Semantic File Inspector_](https://github.com/IS4Code/SFI) is another "anything-to-RDF" tool whose goal is to collect metadata from files and file systems (archives, directories, etc.) using domain-specific vocabularies. You are welcome to try it out (runs completely in the browser) [here](https://sfi.is4.site/#!describe%20-qd%20-x%20analyzer:content-type%20-x%20analyzer:path-object%20-x%20*-hash:*%20-h%20xxh64%20-f%20ttl%20*%20-).

While SFI is focusing more on metadata and external description (for example, for images: dimensions, color depth, EXIF, etc.), as well as identification (making it useful for archival, semantic search, and similar areas), rather than transforming any data fully into RDF, I believe alignment and potential incorporation of the two approaches would be beneficial (SFI runs on .NET, so full interoperability requires IKVM or similar). I have already started incorporating deeper format structure into the output, such as with describing .NET classes, fields, methods, etc. using CodeOntology, and I plan to do the same with XML using the XML Infoset vocabulary. As long as sufficient vocabularies are available, anything is open for addition, and suggestions or contributions are always welcome.

In SFI, the actual content is described at various levels; for data like XML, there could even be 4 in total:
```ttl
# The file as located on disk
_:file
  a nfo:FileDataObject;
  skos:prefLabel "xmlns.xml";
  nfo:fileName "xmlns.xml";
  nfo:fileSize 70 ;
  nfo:fileLastModified "2014-05-17T20:46:50.480000Z"^^xsd:dateTime;
  nie:interpretedAs <ni:///mh;4ucCCNdNlXzirfpf?ct=text/plain>.

# The immediate raw content of the file
<ni:///mh;4ucCCNdNlXzirfpf?ct=text/plain>
  a cnt:ContentAsText;
  skos:prefLabel "text (70 B)"@en;
  dcterms:extent "70"^^dt:byte;
  cnt:characterEncoding "us-ascii";
  at:digest <urn:xxh64:D74D957CE2ADFA5F>;
  dcterms:hasFormat <ni:///mh;4ucCCNdNlXzirfpf?ct=application/xml>.

# The first format which can be parsed from the file
<ni:///mh;4ucCCNdNlXzirfpf?ct=application/xml>
  a schema:MediaObject, cnt:ContentAsXML, xis:Document;
  skos:prefLabel "XML object (elem)"@en;
  xis:documentElement <ni:///mh;4ucCCNdNlXzirfpf?ct=application/xml#element(/1)>;
  cnt:declaredEncoding "utf-8";
  cnt:version "1.0";
  dcterms:hasFormat <ni:///mh;4ucCCNdNlXzirfpf?ct=application/prs.implied-document+xml;root=elem>.

# The media object the XML document corresponds to, implied in this case
<ni:///mh;4ucCCNdNlXzirfpf?ct=application/prs.implied-document+xml;root=elem>
  a schema:MediaObject;
  skos:prefLabel "ELEM object"@en.
```
For an example of possible interoperation, for the last level (i.e. `ni:///mh;4ucCCNdNlXzirfpf?ct=application/prs.implied-document+xml;root=elem`), I plan to use the [`xyz:`](https://sparql.xyz/facade-x/data/) namespace or similar to "interpret" the XML document directly using the elements, attributes etc. inside, rather than describing the XML content/infoset as in the previous level.

**Comment by justin2004** [comment]
i always like to see more X to RDF tools!

though i don't often have a need for metadata -- instead i almost always need the actual content. for example [here](https://github.com/justin2004/weblog/tree/master/git_repo_as_rdf#readme) is something i still find to be useful.

what uses have you found this to help with? got any examples?

**Reply by IS4Code** [reply]
I find the primary uses in archival and file processing.

Large file archives could benefit from having a triple store automatically filled by processing uploaded files, allowing for both semantic search and browsing a file/archive contents without downloading it (see [here](https://archive.data.is4.site/describe/?url=ni:///mh;4ucCCA1uFg5qaoIm?ct=application/octet-stream) for an example) ‒ such a description is usually much smaller than the original file, since only the metadata is preserved (for images, videos, etc.). Moreover, since there is emphasis on using persistent URIs wherever possible, one could even cross-reference datasets for common files or other pieces of data, look for all archives containing a particular file, and similar (check out [this](https://sfi.is4.site/example-queries.html) for example SPARQL queries).

Another possibility would be to use RDF-based tools for tasks like looking for duplicates or changes across file systems, validating files based on the presence/absence of particular file formats or other properties. You could validate any file using SHACL and reject archives containing unknown file formats. 😉
